### ISSUE TITLE
Documentation for supported platforms and specific build instructions?

### ISSUE THREAD

**Issue by FriendFX** [potentialventures/cocotb#164]
I guess I was a bit too optimistic after reading the [Pre-requisites section in the documentation](http://cocotb.readthedocs.org/en/latest/quickstart.html#pre-requisites) which simply mentions

> * Python 2.6+
> * Python-dev packages
> * A Verilog simulator

Having found that cocotb uses Makefiles to run the examples, I thought I could get this to work under my existing Cygwin (64bit, under Win 7) environment, since I had used that in the past to run ModelSim simulations. I was also able to build the latest `iverilog` in this environment and have a Python 2.7 installation *within* Cygwin.

So I installed *additional* requisites (such as `autoconf`, `gcc` and a bunch of others), cloned the "endian_swapper" example and ran `make`. Unfortunately, I couldn't get past the following errors:

> $ make
> make results.xml
> make[1]: Entering directory '/home/friendfx/cocotb/examples/endian_swapper/tests'
> cd ../cosim && make
> make[2]: Entering directory '/home/friendfx/cocotb/examples/endian_swapper/cosim'
> gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -I/usr/include/python2.7 -c io.c -o io_module.o
> gcc: error: unrecognized command line option ‘-fstack-protector-strong’
> Makefile:9: recipe for target 'io_module.o' failed
> make[2]: *** [io_module.o] Error 1
> make[2]: Leaving directory '/home/friendfx/cocotb/examples/endian_swapper/cosim'
> Makefile:65: recipe for target 'hal' failed
> make[1]: [hal] Error 2 (ignored)
> iverilog -o sim_build/sim.vvp -D COCOTB_SIM=1   /home/friendfx/cocotb/examples/endian_swapper/tests/../hdl/endian_swapper.sv
> make -C /home/friendfx/cocotb/lib/embed
> make[2]: Entering directory '/home/friendfx/cocotb/lib/embed'
> gcc -Werror -Wcast-qual -Wcast-align -Wwrite-strings -Waggregate-return -Wstrict-prototypes -Wall -Wno-unused-parameter -fno-common -g -DDEBUG -fpic -DPYTHON_SO_LIB=libpython2.7.so -shared -Xlinker -export-dynamic -L/usr/lib -o /home/friendfx/cocotb/build/libs/x86_64/libcocotb.so /home/friendfx/cocotb/build/obj/x86_64/gpi_embed.o -ldl -lpython2.7 -L/home/friendfx/cocotb/build/libs/x86_64
> /usr/lib/gcc/x86_64-pc-cygwin/4.8.3/../../../../x86_64-pc-cygwin/bin/ld: warning: --export-dynamic is not supported for PE+ targets, did you mean --export-all-symbols?
> /home/friendfx/cocotb/build/obj/x86_64/gpi_embed.o: In function `embed_sim_init':
> /home/friendfx/cocotb/lib/embed/gpi_embed.c:126: undefined reference to `gpi_get_root_handle'
> /home/friendfx/cocotb/lib/embed/gpi_embed.c:126:(.text+0x12b): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `gpi_get_root_handle'

It was only then that I came across the discussion in [Issue #144](https://github.com/potentialventures/cocotb/issues/144) which asks for "Windows Makefile support" and looks quite a bit more daunting to me than the "Pre-requisites" in the documentation had suggested.

Long story short, is there a document somewhere which platforms are supported or cocotb has been tested with? Additionally, could someone point me to the "method of compiling/linking against the 32-bit libraries on a 64-bit Linux platform" mentioned in [Issue #144](https://github.com/potentialventures/cocotb/issues/144), and something equivalent for other environments?

From that discussion, my questions would include:
 * Do I have to use a 32 bit platform?
 * Do I need to use specific Python or simulator versions?
 * What simulator-specific pre-requisites do I need?
 * Which environment variables do I need to set/modify so the build system can find everything it needs?

**Comment by stuarthodgson** [comment]
@jmorarity

Regarding your segfault, we have just fixed one with modelsim on Branch Issue#188. Not sure if this is what you where seeing though. Could test? on whatever version you have?

**Comment by jmorarity** [comment]
No change for me on 10.2c or 10.3d.  Still builds but seg faults.

>vsim -c -do runsim.do 2>&1 | tee sim.log
>Reading C:/Mentor/modelsim_dlx_10.3d/tcl/vsim/pref.tcl

>\# 10.3d

>\# do runsim.do
>\# Model Technology ModelSim DE vlog 10.3d Compiler 2014.10 Oct  7 2014
>\# Start time: 21:53:46 on Jan 15,2015
>\# vlog -timescale 1ns/100ps -mfcu "+acc=rmb" -sv c:/repos/pycosim_trunk/cocotb_tst2/src/examples/endian_swapper/tests/../hdl/endian_swapper.sv
>\# -- Compiling module endian_swapper_sv
>\#
>\# Top level modules:
>\#       endian_swapper_sv
>\# End time: 21:53:46 on Jan 15,2015, Elapsed time: 0:00:00
>\# Errors: 0, Warnings: 0
>\# Attempting stack trace sig 11
>\# Signal caught: signo [11]

>\# PARTIAL STACK TRACE ONLY, anomaly 2
>\# vsim_stacktrace.vstf written
>\# Current time Thu Jan 15 21:53:47 2015
>\# ModelSim Stack Trace
>\# Program = vsim
>\# Id = "10.3d"
>\# Version = "2014.10"
>\# Date = "Oct  7 2014"
>\# Platform = win32pe
>\# --> START OF USERCODE
>\# 0    0x6fc89111: 'ZNSsC1EPKcRKSaIcE + 0x39' in 'c:\python27\gnuplot\binary\libstdc++-6.dll'
>\# 1    0x66681ee9: 'handle_vpi_callback + 0x178' in 'c:\repos\pycosim_trunk\cocotb_tst2\src\build\libs\i686\libvpi.dll'
>\# <-- END OF USERCODE
>\# 2    0x00575e06: '<unknown (@0x575e06)>'
>\# 3    0x00575fdb: '<unknown (@0x575fdb)>'
>\# 4    0x0057637b: '<unknown (@0x57637b)>'
>\# 5    0x006b3045: '<unknown (@0x6b3045)>'
>\# 6    0x006b4b99: '<unknown (@0x6b4b99)>'
>\# 7    0x009bdfc2: '<unknown (@0x9bdfc2)>'
>\# 8    0x1000cb1f: 'TclInvokeStringCommand + 0x8f' in 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
>\# 9    0x100473b5: 'Tcl_ExprObj + 0x1605' in 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
>\# 10   0x100462b5: 'Tcl_ExprObj + 0x505' in 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
>\# 11   0x1000f6a9: 'TclEvalObjEx + 0x2f9' in 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
>\# End of Stack Trace

>\# vsim -c -do "runsim.do"
>\# Start time: 21:53:46 on Jan 15,2015
>\# Loading libvpi.dll
>make[1]: Leaving directory `/c/repos/pycosim_trunk/cocotb_tst2/src/examples/endian_swapper/tests'

**Comment by stuarthodgson** [comment]
This is a bit strange

# 0 0x6fc89111: 'ZNSsC1EPKcRKSaIcE + 0x39' in
'c:\python27\gnuplot\binary\libstdc++-6.dll'
# 1 0x66681ee9: 'handle_vpi_callback + 0x178' in
'c:\repos\pycosim_trunk\cocotb_tst2\src\build\libs\i686\libvpi.dll'

It looks from this that your build is picking up the stdc++ lib from your
python gnuplot installation rather than from mingw.

Can you modify the PATH to place <mingw_install>/bin ahead of the python
installation?

make clean
make

Stu

**Comment by jmorarity** [comment]
Ok, So that strange error was indeed the wrong "libstdc++-6.dll" being
pulled (from Python extras rather than MinGW).  It's a bit concerning how
dependent it is on the runtime path order though.  A search of my computer
brings up 10 different locations for "libstdc++-6.dll" .. everything from
MinGW and Python to Xilinx and Mathworks tools.  Would it not be better to
statically link it, so the proper version is at least bound to the Cocotb
DLL after being built?

The good news of course is that I'm up and running with the latest
version!  I still had to make a couple of small edits to build and run my
custom sims. I'm going to attempt to put together a pull request for those
items ... but bear with me, I'm a GitHub virgin :)

**Comment by potentialventures** [comment]
On 18 Jan 2015 07:04, "J.Morarity" <notifications@github.com> wrote:
>
> Ok, So that strange error was indeed the wrong "libstdc++-6.dll" being
> pulled (from Python extras rather than MinGW). It's a bit concerning how
> dependent it is on the runtime path order though.

Well, not really. A valid question is wy does python bring in it's own libc.

A search of my computer
> brings up 10 different locations for "libstdc++-6.dll" .. everything from
> MinGW and Python to Xilinx and Mathworks tools. Would it not be better to
> statically link it, so the proper version is at least bound to the Cocotb
> DLL after being built?

I'll  need to look into how windows works in that regard. On Linux its just
put the location of the lib in the rpath. Windows is different though.

>
> The good news of course is that I'm up and running with the latest
> version!

This is great. Now means we can pull documentation together and do a
release soon.

I still had to make a couple of small edits to build and run my
> custom sims. I'm going to attempt to put together a pull request for
those
> items ... but bear with me, I'm a GitHub virgin :)

Bear with us we are windows virgins.

Stu
>
> On Fri Jan 16 2015 at 1:03:56 AM Stuart Hodgson <notifications@github.com>

> wrote:
>
> > This is a bit strange
> >
> > # 0 0x6fc89111: 'ZNSsC1EPKcRKSaIcE + 0x39' in
> > 'c:\python27\gnuplot\binary\libstdc++-6.dll'
> > # 1 0x66681ee9: 'handle_vpi_callback + 0x178' in
> > 'c:\repos\pycosim_trunk\cocotb_tst2\src\build\libs\i686\libvpi.dll'
> >
> > It looks from this that your build is picking up the stdc++ lib from
your
> > python gnuplot installation rather than from mingw.
> >
> > Can you modify the PATH to place <mingw_install>/bin ahead of the
python
> > installation?
> >
> > make clean
> > make
> >
> > Stu
> >
> > On 16 January 2015 at 06:04, J.Morarity <notifications@github.com>
wrote:
> >
> > > No change for me on 10.2c or 10.3d. Still builds but seg faults.
> > >
> > > vsim -c -do runsim.do 2>&1 | tee sim.log
> > > Reading C:/Mentor/modelsim_dlx_10.3d/tcl/vsim/pref.tcl
> > >
> > > # 10.3d
> > >
> > > # do runsim.do
> > > # Model Technology ModelSim DE vlog 10.3d Compiler 2014.10 Oct 7 2014
> > > # Start time: 21:53:46 on Jan 15,2015
> > > # vlog -timescale 1ns/100ps -mfcu "+acc=rmb" -sv
> > >
> >
c:/repos/pycosim_trunk/cocotb_tst2/src/examples/endian_swapper/tests/../hdl/

> >
> > > endian_swapper.sv
> > > # -- Compiling module endian_swapper_sv
> > > #
> > > # Top level modules:
> > > # endian_swapper_sv
> > > # End time: 21:53:46 on Jan 15,2015, Elapsed time: 0:00:00
> > > # Errors: 0, Warnings: 0
> > > # Attempting stack trace sig 11
> > > # Signal caught: signo [11]
> > >
> > > # PARTIAL STACK TRACE ONLY, anomaly 2
> > > # vsim_stacktrace.vstf written
> > > # Current time Thu Jan 15 21:53:47 2015
> > > # ModelSim Stack Trace
> > > # Program = vsim
> > > # Id = "10.3d"
> > > # Version = "2014.10"
> > > # Date = "Oct 7 2014"
> > > # Platform = win32pe
> > > # --> START OF USERCODE
> > > # 0 0x6fc89111: 'ZNSsC1EPKcRKSaIcE + 0x39' in
> > > 'c:\python27\gnuplot\binary\libstdc++-6.dll'
> > > # 1 0x66681ee9: 'handle_vpi_callback + 0x178' in
> > > 'c:\repos\pycosim_trunk\cocotb_tst2\src\build\libs\i686\libvpi.dll'
> > > # <-- END OF USERCODE
> > > # 2 0x00575e06: ''
> > > # 3 0x00575fdb: ''
> > > # 4 0x0057637b: ''
> > > # 5 0x006b3045: ''
> > > # 6 0x006b4b99: ''
> > > # 7 0x009bdfc2: ''
> > > # 8 0x1000cb1f: 'TclInvokeStringCommand + 0x8f' in
> > > 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
> > > # 9 0x100473b5: 'Tcl_ExprObj + 0x1605' in
> > > 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
> > > # 10 0x100462b5: 'Tcl_ExprObj + 0x505' in
> > > 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
> > > # 11 0x1000f6a9: 'TclEvalObjEx + 0x2f9' in
> > > 'c:\mentor\modelsim_dlx_10.3d\win32pe\tcl85.dll'
> > > # End of Stack Trace
> > >
> > > # vsim -c -do "runsim.do"
> > > # Start time: 21:53:46 on Jan 15,2015
> > > # Loading libvpi.dll
> > > make[1]: Leaving directory
> > >
`/c/repos/pycosim_trunk/cocotb_tst2/src/examples/endian_swapper/tests'
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
https://github.com/potentialventures/cocotb/issues/164#issuecomment-70212204>

> >
> > > .
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
https://github.com/potentialventures/cocotb/issues/164#issuecomment-70225829>
